Proposal: add hasCrossSiteAncestor value to partitionKey in Cookies API (#581)
hober/web-extensions: proposals/hasCrossSiteAncestor.md

# Proposal: Update Extensions to interact with Partitioned Cookies containing Cross-Site Ancestor Chain Bit values

**Summary**

Allow extensions to utilize a hasCrossSiteAncestor boolean value when interacting with partitioned cookies that include a cross-site ancestor chain bit in their partiton key.

> #### Brackground information: Description of Cross-site Ancestor Chain Bit in partitioned cookies

Reviewer: ```suggestion
#### Background information: Description of Cross-site Ancestor Chain Bit in partitioned cookies
```

> A cross-site ancestor chain bit is a value that is set by the browser when a paritioned cookie is created. The value is used as part of the key in the partition key to determine access to a partitioned cookie. If the bit indicates true, if the cookie has been set in a third third party context.  

Reviewer: Fixing a typo and also trying to clarify the explanation a bit:

```suggestion
A cross-site ancestor chain bit is a value that is set by the browser when a partitioned cookie is created. The value is used as part of the key in the partition key to determine access to a partitioned cookie. If the bit indicates true, it means the cookie has been set in a third party context.
```

A third party context occurs when any parent frame, in the frame tree, is cross-site with the frame the cookie is being set on. If one site of the frame tree is third party, then all children frames of that frame are also third party and have a cross site ancestor.

> | Site frame tree |Site cookie is set on| hasCrossSiteAncestor value of cookie|

Reviewer: Can we explicitly state that this example defines anything starting with A (A1, A2, A3) to be first-party with each other? I think that is not immediately clear right now to someone new to this.

|---|---|--|
| A1         |A1| false         | 
| A1->A2      |A2 | false          | 
| A1->A2->A3   |A3   | false          | 
| A1->B       |B| true          |
| A1->B->A2    |A1   | false          |
| A1->B->A2 |B    | true          |
| A1->B->A2 |A2   | true          |

**Document Metadata**

> **Author:**  aselya 

Reviewer: ```suggestion
**Author:** aselya 
```

**Sponsoring Browser:** Chrome

**Contributors:** DCtheTall

**Created:** 2024-04-01

**Related Issues:** [PrivacyCG/CHIPS issue 6 - How do Partitioned cookies interact with browser extensions?](https://github.com/privacycg/CHIPS/issues/6)

## Motivation
https://github.com/privacycg/CHIPS/issues/40 is adding a cross-site ancestor bit value to partitioned cookies.

### Objective
To interact with partitioned cookies containing a cross-site ancestor chain bit correctly, extensions will need to have the ability to to specify a value (hasCrossSiteAncestor) that corresponds to the value of the cross-site ancestor chain bit in partitioned cookies.

#### Use Cases

#### Password Manager:

> Let’s say a password manager extension (with host permissions) is used by users to access their login information by setting a cookie that stores their usernames and passwords in an encrypted partitioned cookie. To protect their users against clickjacking, the extension adds a setting that prevents their cookies from being accessed, by default, in embeds that have cross site ancestors without triggering a user prompt. If permission is given through the prompt, the extension sets a cookie with a hasCrossSiteAncestor value of true. Upon subsequent visits, the extension checks the cookie store for the presence of a cookie with a hasCrossSiteAncestor value of true to determine whether the prompt needs to be rendered.

Reviewer: Hmm, I didn't find this example particularly convincing:

- Is the extension storing the cookies on a chrome-extension:// page? If so, that wouldn't work because chrome-extension:// schemes aren't considered "secure" for cookie purposes and therefore you aren't able to set the `Partitioned` attribute.
- Alternatively, is the suggestion a password manager would use encrypted cookies on the origin you are signing in to? I'm not sure that's a realistic path we would encourage developers down.

I wonder if we could go with something simpler like a cookie manager extension that wants to give users full insight into all of the cookies that might be stored.

Reviewer 2: I think that in this use case, the "password manager" extension uses cookies associated with a third party website as a storage mechanism. Odd use case but I find it useful to illustrate why this could be desirable for an extension API to have this functionality.

A cookie manager extension would be an obvious use case, and the justification can be as simple as needing a way to uniquely identify specific cookies in order to update or remove them.

Author: I don't think there's any harm in including two use cases. I'll add @oliverdunk's example to the PR.

To allow for this protection and UX flow to work, the extension would need to have the ability to set/get cookies with specific hasCrossSiteAncestor values. 

### Known Consumers
All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.

## Specification

### Schema

Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

```

> //Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.

Reviewer: ```suggestion
// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
```

> "details" {

Reviewer: ```suggestion
"details": {
```

  "partitionKey": {

>     "topLevelSite" : "https://example.com",

Reviewer: ```suggestion
    "topLevelSite": "https://example.com",
```

>     "topLevelSite" : "https://example.com"

Reviewer: ```suggestion
    "topLevelSite" : "https://example.com",
```

>     "hasCrossSiteAncestor" : true

Reviewer: ```suggestion
    "hasCrossSiteAncestor": true
```

# Proposal: Update Extensions to interact with Partitioned Cookies containing Cross-Site Ancestor Chain Bit values

**Summary**

Allow extensions to utilize a hasCrossSiteAncestor boolean value when interacting with partitioned cookies that include a cross-site ancestor chain bit in their partiton key.

#### Background information: Description of Cross-site Ancestor Chain Bit in partitioned cookies

> A cross-site ancestor chain bit is a value that is set by the browser when a partitioned cookie is created. The value is used as part of the key in the partition key to determine access to a partitioned cookie. If the bit indicates true, it means the cookie has been set in a third party context.

Reviewer 3: First two sentences could be combined:

"The cross-site ancestor chain bit is a component of the cookie partition key that is set by the browser when keys are created."

> A cross-site ancestor chain bit is a value that is set by the browser when a partitioned cookie is created. The value is used as part of the key in the partition key to determine access to a partitioned cookie. If the bit indicates true, it means the cookie has been set in a third party context.

Reviewer 3: Nit: "first-" and "third-party" are hyphenated when used as an adjective.

> A third party context occurs when any parent frame, in the frame tree, is cross-site with the frame the cookie is being set on. If one site of the frame tree is third party, then all children frames of that frame are also third party and have a cross site ancestor. 

Reviewer 2: ```suggestion
A third party context occurs when any parent frame, in the frame tree, is cross-site with the frame or subresource the cookie is being set on. If one site of the frame tree is third party, then all children frames of that frame are also third party and have a cross site ancestor. 
```

Suggestion, provided that it is accurate: frame -> frame or subresource.

Previously the description gave the impression that only the frame tree is considered in determining the ancestor bit. But a subresource (image, style, script, etc) can also have a meaningful answer to "are we talking about cross-site cookies?"

Author: That's a good clarification.

> A third party context occurs when any parent frame, in the frame tree, is cross-site with the frame or subresource the cookie is being set on. If one site of the frame tree is third party, then all children frames of that frame are also third party and have a cross site ancestor. 

Reviewer 3: This is not completely comprehensive. A few cases when the partition key may be cross-site not included here are:

- When the top-level frame makes a 3P subresource request
- When a 1P request is redirected to a 3P URL (even if it is subsequently redirected back to a 1P request after, that subsequent 1P request is now considered an ABA request)

> A third-party context occurs when any parent frame, in the frame tree, is cross-site with the frame or subresource the cookie is being set on. If one site of the frame tree is third-party, then all children frames of that frame are also third-party and have a cross site ancestor. Third-party contexts also occur when the top-level frame makes a third-party subresource request and when a first-party request is redirected to a thrid-party URL (even if it is subsequently redirected back to a first-party request after, that subsequent first-party request is now considered an ABA request).

Reviewer 2: As I pondered on edge cases, I realized that there is no clear definition of what CSA is. I tried to come up with a precise definition, but ran into ambiguity. This whole comment has nothing to do with the extension API, but with defining the primitive (CSA) that is being exposed through the extension APIs.

I quoted the full current text in order to review specific bits. I didn't modify anything within the quoted parts.

> A third-party context occurs when any parent frame, in the frame tree, is cross-site with the frame or subresource the cookie is being set on.

This first line (and the table) gives the impression that the CSA bit is only affected by the frame tree. Later in the paragraph, you do however also mention redirects. For clarity, let's lead with an all-encompassing definition, or add "primarily" to emphasize that although most common there may be other situations.

How about navigations? A third-party child frame navigates the top frame - is the result still third-party with hCSA set?

> If one site of the frame tree is third-party, then all children frames of that frame are also third-party and have a cross site ancestor.

This part is not strictly needed in the definition because it automatically follows from the definition. I would include it after the definition since it does help with the understanding.

> Third-party contexts also occur when the top-level frame makes a third-party subresource request and when a first-party request is redirected to a thrid-party URL

This definition limits itself to subresource requests initiated from the top frame. I assume that the redirect logic also applies to subresource requests from same-site iframes. And what about navigations in general? A child frame could open a popup, and if they are related browsing contexts, I would fully expect them to inherit the topLevelSite + CSA bit.

An iframe is also a subresource. When the redirect chain for an iframe request is ABA, resulting in an AA frame tree, is that child frame third-party? If not, use "non-document subresource" to avoid ambiguity.

> (even if it is subsequently redirected back to a first-party request after, that subsequent first-party request is now considered an ABA request).

This also follows from the definition. And again this is worth mentioning.

Here is a suggested edit that has a precise definition. Rephrase as needed to make it accurate.

```suggestion
A third-party context occurs when the subresource the cookie is being set on has a cross-site frame ancestor, opener or redirect entry.
Once a frame is considered to be in a third-party context, all requests within that frame and its child frames are also third-party and have a cross site ancestor. Similarly, once a request redirects to a cross-site URL, it is considered third-party (even if it is subsequently redirected back to a first-party request after, that subsequent first-party request is now considered an ABA request).
```

Author: Due to community [feedback](https://github.com/privacycg/CHIPS/issues/40#issuecomment-2178846023) and compatibility issues, a redirect and opener are not being considered cross-site in the chromium implementation. I'm fine with the language @Rob--W, proposed as it is correct but I wanted to get some opinion on if it makes sense to alter the language to reflect the implementation or keep it as is. 

@oliverdunk, @DCtheTall

Reviewer 4: > A third-party context occurs when the subresource the cookie is being set on has ...

How is this applied to a cross-site fetch request whose response sets a partitioned cookie? What's the "subresource" in this case?

Author: @HeikoTheissen, in the case of extensions, there isn't a subresource. 

Since the ancestor chain bit is a new element of the cookie partition key, this language was added to provide developers, who don't have prior knowledge of the ancestor chain bit, a definition to help understand it's purpose.

Reviewer 2: > Due to community [feedback](https://github.com/privacycg/CHIPS/issues/40#issuecomment-2178846023) and compatibility issues, a redirect and opener are not being considered cross-site in the chromium implementation. I'm fine with the language @Rob--W, proposed as it is correct but I wanted to get some opinion on if it makes sense to alter the language to reflect the implementation or keep it as is.

Adjust the phrasing to match the implementation / final design. The purpose of my suggested language was to clarify the actual behavior, and that should closely match the final design. If there are parts that may change, it would also be fine to call out that the behavior may change as the implementation evolves and the spec matures.

If you are in favor of the suggested text, providing review feedback may be easier if you apply the suggested text. Then other reviewers can suggest new improvements to the (applied) suggestion instead of trying to somehow read the diff from my comment.

> > > A third-party context occurs when the subresource the cookie is being set on has ...
> >
> > How is this applied to a cross-site fetch request whose response sets a partitioned cookie? What's the "subresource" in this case?
>
> @HeikoTheissen, in the case of extensions, there isn't a subresource.

This response puzzles me. By the term "subresource", I meant all non-document requests. What meaning of "subresource" are you using? Could you elaborate what you mean by "in the case of extensions, there isn't a subresource"?

Reviewer 4: > By the term "subresource", I meant all non-document requests.

That makes sense. (I had initially thought of the subresource as a frame, image or something in the document tree.)

Author: Thank you for the suggestion, I've applied the text.

I should have been more specific with my language as upon reading it again it's clear I misrepresented my intent. 

In the context of the issue that, @HeikoTheissen, we had been discussing redirect requests from a subresource and how the browser calculates the value of the ACB. The issue that was encountered, wouldn't impact this API not because there is no subresource but because the API has the ability to set the value of the ACB directly which webpages do not.

Note: In the table below, sites A1, A2 and A3 are all first party to each other.
| Site frame tree |Site cookie is set on| hasCrossSiteAncestor value of cookie|
|---|---|--|
| A1         |A1| false         | 
| A1->A2      |A2 | false          | 
| A1->A2->A3   |A3   | false          | 
| A1->B       |B| true          |
| A1->B->A2    |A1   | false          |
| A1->B->A2 |B    | true          |
| A1->B->A2 |A2   | true          |

**Document Metadata**

> **Author:** aselya 

Reviewer 3: Optional, I think it's nice to make these author credits include links to our profiles :)

> | A1->B->A2    |A1   | false          |

Reviewer 2: ```suggestion
| A1->B->A2    |A1   | true          |
```

The first column reflects the full frame tree, the second column the domain of a (sub)resource in the deepest frame for which a cookie is managed. Since the frame tree is ABA, the bit should be true instead of false.

Author: Made and update to the table to help clarify what nodes have access and what the cross-site ancestor value should be.

Author: @Rob--W, does my clarification make sense?

### Additional Security Considerations
None

## Alternatives
The hasCrossSiteAncestor value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the topLevelSite in the partitionKey. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for hasCrossSiteAncestor since it would not be explicitly provided by the extension.

### Existing Workarounds
To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.

### Open Web API
The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.

## Implementation Notes
APIs affected by the change and the behavior assoicated with the change:

- `cookies.get()`:

> If there is a partitionKey present and it has a value for topLevelSite but no value for hasCrossSiteAncestor, the get method will try to deduce what the value for hasCrossSiteAncestor is likely to be when getting the cookie. In the event that the value is incorrect, developers can either pass the correct value for hasCrossSiteAncestor or use cookies.getAll() with no hasCrossSiteAncestor value set.

Reviewer 2: For consistency with set/remove, throw if hCSA is set without topLevelSite. Assuming that hCSA is meaningless without topLevelSite.

Author: Sounds good to me

- `cookies.getAll()`: 
If no value is set for hasCrossSiteAncestor cookies with both true and false values for hasCrossSiteAncestor will be returned. Otherwise, cookies will be returned that match the topLevelSite and the passed value for hasCrossSiteAncestor.

- `cookies.set()`: 

> As described the Abuse Mitigations section, this method will not allow a hasCrossSiteAncestor value of false, if the URL associated with the cookie and the topLevelSite in the partitionKey are not first-party. If this is attempted, an error will be returned. If no topLevelSite is provided and a hasCrossSiteAncestor value is provided, the cookie will not be set and an error will be returned.

Reviewer 2: Can set/remove have the same deduction logic for hasCrossSiteAncestor as get()? Consistent behavior ensures that extensions that somehow construct partitionKey manually without hCSA will operate within the same partition.

Author: That sounds good to me.

All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.

## Specification

### Schema

Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

```
// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
"details": {
  "partitionKey": {
    "topLevelSite": "https://example.com",
    "hasCrossSiteAncestor": true
  }
}
```

### New Permissions
No new permissions are required.

### Manifest File Changes
No new manifest fields are required.

## Security and Privacy
Privacy: The hasCrossSiteAncestor boolean value does not reveal private information.

Security: The hasCrossSiteAncestor boolean value does not impact security.

### Exposed Sensitive Data
No sensitive data is associated with the cross-site ancestor chain bit.

### Abuse Mitigations

> A cookie may only have no cross-site ancestor when the topLevelSite in the partitionKey and the URL with which the cookie is associated with are first-party to each other. To prevent the creation of cookies that violate this, the set method will return an error if the URL and the topLevelSite are not first-party for cookies that are set with no-cross site ancestor.

Reviewer 2: When the public suffix list is updated, it is possible for the meaning of "first-party" (derived from eTLD+1) to change.

For a cookie manager type of add-on, it is important that they should be able to migrate or at least remove data even if the combination is invalid. The current proposal does not restrict parameters to get, getAll, remove, which enables this use case. The set method is constrained, but I think that it's an acceptable trade-off between the edge case vs general utility.

Would it make sense to capture this? From the proposal it currently seems that the hCSA bit is static and fully well-defined, but the possibility of "first-party" changing at runtime (public suffix list update) makes it a bit more complicated, and is relevant to the design of the API.

Author: Added language to capture this explicitly.

### Existing Workarounds
To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.

### Open Web API
The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.

## Implementation Notes

### Populating hasCrossSiteAncestor when not provided
When no value has been provided for `hasCrossSiteAncestor`, if the `domain` associated with the `cookie` is same-site to the value of the `topLevelSite`, the `hasCrossSiteAncestor` value will be set to false otherwise the value will be set to true.

### APIs affected by the change and the behavior assoicated with the change:

- `cookies.get()`:

> If no `hasCrossSieAncestor` value is provided it will be populated using the algorithim described above. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.

Reviewer 2: ```suggestion
If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithim described above. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.
```

# Proposal: Update Extensions to interact with Partitioned Cookies containing Cross-Site Ancestor Chain Bit values

**Summary**

Allow extensions to utilize a hasCrossSiteAncestor boolean value when interacting with partitioned cookies that include a cross-site ancestor chain bit in their partiton key.

#### Background information: Description of Cross-site Ancestor Chain Bit in partitioned cookies

> The cross-site ancestor chain bit is a component of the cookie partition key that is set by the browser when keys are created. If the bit indicates true, it means the cookie has been set in a third-party context.

Reviewer 2: (continuation of thread from https://github.com/w3c/webextensions/pull/581#discussion_r1574385184)

Questions that I have that are not answered by this section:

- (When) can top-level documents have the bit?
- (When) can unpartitioned cookies have the bit?

Top-level navigations triggered from iframes (whether `_top` or a new popup) are scenarios where these questions become relevant.

Author: >(When) can top-level documents have the bit?

In the chromium implementation, unpartitioned cookies do not have the bit because unpartitioned cookies do not have a `partitionKey`.  In the current implementation, passing an empty value for `topLevelSite` indicates an unpartitioned cookie. To maintain backwards compatibility, I think it makes sense to have the api ignore the value of `hasCrossSiteAncestor` when the `topLevelSite` is empty. If you're ok with this I can update the language of the proposal to reflect it.

>(When) can top-level documents have the bit?

Regardless of the source of the navigation, top-level documents will always have a  `hasCrossSiteAncestor` value of `false`, indicating no cross-site ancestor, since as the top-level they do not have any ancestor. This choice was made as a result of compatibility issues detailed in this [bug](https://g-issues.chromium.org/issues/339410000).

Reviewer 2: Thanks for your clarification. The "Unpartitioned cookies do not have the bit" part was not obvious to me, and it would be worth calling that characteristic explicitly in the proposal document. "Unpartitioned cookies always have hasCrossSiteAncestor false."

> > (When) can top-level documents have the bit?
> 
> In the chromium implementation, unpartitioned cookies do not have the bit because unpartitioned cookies do not have a `partitionKey`. In the current implementation, passing an empty value for `topLevelSite` indicates an unpartitioned cookie. To maintain backwards compatibility, I think it makes sense to have the api ignore the value of `hasCrossSiteAncestor` when the `topLevelSite` is empty. If you're ok with this I can update the language of the proposal to reflect it.

Why do you think that we should treat `hasCrossSiteAncestor` specially when `topLevelSite` is empty? I can't think of a benefit for extensions or browsers to special case this, can you?

Author: >Why do you think that we should treat hasCrossSiteAncestor specially when topLevelSite is empty? I can't think of a benefit for extensions or browsers to special case this, can you?

How about adding language saying, "Unpartitioned cookies, which are indicated by a cookie not containing a `partitionKey`, the key not containing a `topLevelSite` or an empty `topLevelSite`, always have a `hasCrossSiteAncestor` value of `false`.

This addresses how the value should be handled when the cookie is unpartitioned and clarifies when a cookie is considered unpartitioned.

Reviewer 2: The proposed language SGTM. It clarifies the behavior with unpartitioned cookies.

The situation with top-level documents is still under-specified in this document. Can you also add a note about top-level documents? Based on your reply in the other thread at https://github.com/w3c/webextensions/pull/581#discussion_r1707027673, I think that Chrome can have partitioned cookies in top-level documents.

"Top-level documents may have partitioned cookies (topLevelSite non-empty, i.e. set to the site of the top-level document), but hasCrossSiteAncestor is always false for top-level contexts by design (CHIPS spec link here)."

This is mainly to help extension developers with interpreting the behavior correctly, independent of browser-specific behavior.

In Firefox, top-level documents are unpartitioned, and therefore hasCrossSiteAncestor is false (following the reasoning earlier in this thread). Third-party frames (including frames that are same-site to the top document but with a cross-site frame in between, but excluding frames whose full ancestor chain is same-site) are partitioned automatically. Third-party frames can opt in to unpartitioned cookies with `requestStorageAccess`.

In Chrome, partitioning seems to be tied to the `Partitioned` cookie attribute that the website has pass to `Set-Cookie`, instead of forced by the browser. While Firefox does not partition cookies in top-level documents, it seems that in Chrome a top-level document can still have partitioned cookies, i.e. `topLevelSite` being non-empty, if the cookie is set with the `Partitioned` attribute.

(If my understanding is wrong, please correct me - I am spelling out my thoughts process to make it easier to challenge my misconceptions if any.)

Author: You have the correct understanding of the differences between the implementations.

I like the language you suggested as well and will add it.

A third-party context occurs when the subresource the cookie is being set on has a cross-site frame ancestor.
Once a frame is considered to be in a third-party context, all requests within that frame and its child frames are also third-party and have a cross site ancestor. Similarly, once a request redirects to a cross-site URL, it is considered third-party (even if it is subsequently redirected back to a first-party request after, that subsequent first-party request is now considered an ABA request).

Note: In the table below, sites A1, A2 and A3 are all first-party to each other.
| Site frame tree |Site cookie is set on| hasCrossSiteAncestor value of cookie| Nodes that can't access the cookie|
|---|---|--|--|
| A1         |A1| false         | |
| A1->A2      |A2 | false          | 
| A1->A2->A3   |A3   | false          |  
| A1->B       |B| true          | B|

> | A1->B->A2    |A1   | false        | B and A2
> | A1->B->A2 |B    | true          | A1 and A2 |

Reviewer 2: (continued from original thread at https://github.com/w3c/webextensions/pull/581#discussion_r1651549877)

Initially I assumed that "the second column the domain of a (sub)resource in the deepest frame for which a cookie is managed".
Your clarification suggests that the actual meaning is actually, "the second column the site of the frame in the frame tree where the cookie is set".

Since the descendants of a frame have zero effect on the ancestor bit of a cookie, the behavior of `A1->B->A2 | A1` is equivalent to `A1 | A1` (already at line 17), and `A1->B->A2 | B` is equivalent to `A1->B | B` (line 20). Is there any benefit to listing an example with descendants if descendants don't impact the behavior of the cookies in upper frames?

```suggestion
```

Author: I'm fine removing it.

To allow for this protection and UX flow to work, the extension would need to have the ability to set/get cookies with specific hasCrossSiteAncestor values. 

### Known Consumers
All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.

## Specification

### Schema

Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

```
// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
"details": {
  "partitionKey": {
    "topLevelSite": "https://example.com",
    "hasCrossSiteAncestor": true
  }
}
```

### New Permissions
No new permissions are required.

### Manifest File Changes
No new manifest fields are required.

## Security and Privacy

> Privacy: The hasCrossSiteAncestor boolean value does not reveal private information.

Reviewer 2: Privacy risk: "incorrect" use of the `cookies` API can result in cookies being inadvertently associated with the wrong cookie partition. See my comment at `getAll`.

Regardless of whether that issue is fixed, we need to call out the risk of inadvertent mixing cookies of cookies from different partitions by extensions that are unaware of partitioning.

Note that it is currently near-impossible for an extension to "correctly" use the `cookies` API in combination with actual web pages. For the simple use case of "create cookie for this tab" or "remove cookie for this tab" ("this tab" = top-level document), the documented examples simply take the URL of the current tab to feed to the cookies API:

- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/set#examples
- https://github.com/GoogleChrome/chrome-extensions-samples/blob/f33eb566803f6c98786efde7232a688d67211d2b/_archive/mv2/api/cookies/manager.js#L131-L134

With the introduction of the cross-site ancestor bit, that logic is unreliable. At best it removes/edits the right cookie. It is also probably for it to not remove/modify any cookie. At worst it removes/modifies the wrong cookie.

At the bare minimum, we'd need a way to query the hasCrossSiteAncestor bit for a given tab (tabId), frame (frameId) and/or document (documentId). This can be a dedicated `cookies.getHasCrossSiteAncestor({ tabId, frameId, documentId })`.

I also considered a generic `cookies.getPartitionKey` method (which also takes a URL in order to determine whether the cookie would be cross-site or same-site). The advantage of this is that extension developers can then treat the return value as a black box and pass it to the `cookies` API to always correctly manage the desired cookie, and that the method can account for `requestStorageAccess` calls in the document if desired.

Author: I agree that it makes sense to provide a way to help developers obtain the expected `HasCrossSiteAncestor` value. I'd be in support of a `cookies.getPartitionKey` method and don't mind drafting a proposal for it.

Author: Added language that describes the preexisting risk, in the privacy risk

Reviewer 2: Are you going to add getPartitionKey to this proposal or would it be a separate document?

On the one hand it is not strictly necessary for extensions that only use the cookies API to get a view of the partitioned cookies. On the other hand it would be nice if the introduction of the concept of "cross-site ancestor chain bit" also includes a first-class method for extensions to accurately determine the correct bit to use, so they can update their extension and immediately use the right mechanism from the start.

Author: I don't have a preference, is there one the you prefer over the other?

Reviewer 2: Let's add it to this proposal then, since the motivation for the new method is directly tied to the feature proposed here.

Author: SGTM. One question that I have is how this API should interact with the [Storage Access API](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API).

In chromium, SAA only impacts unpartitioned cookies. Since unpartitioned cookies in chromium don't have a partitionKey, the default value of false would not change.

In the firefox implementation, is the partitionKey that is created dependent on if a frame has been granted storage access? 
If so, how should we address a case where a cross-site iframe is embedded in a site and`cookies.getHasCrossSiteAncestor` is called. Then the frame separately is granted SAA which would cause the key to change.

Reviewer 2: How it works internally in Firefox is explained at https://searchfox.org/mozilla-central/rev/c4966e1c1e6a8cb38c15a27693e8ecd63082055e/toolkit/components/antitracking/StoragePrincipalHelper.h#15

The short version is that a document may initially have access to partitioned storage. Then after granted access through the SSA, it may from then on have access to unpartitioned cookies, and no access to partitioned cookies.

> If so, how should we address a case where a cross-site iframe is embedded in a site and `cookies.getHasCrossSiteAncestor` is called. Then the frame separately is granted SAA which would cause the key to change.

Good question. From the extension API perspective, I would expect "whatever the effective access" is, and that is something that you should raise in the Privacy CG.

Even without SSA, I have already observed that with CHIPS through the Partitioned attribute, that it is possible for the top-level document to have cookies from two partitions. So perhaps we should shape the new getPartitionKey API such that it recognizes that there are multiple partition keys, and return all of them that may potentially appear, with the default one clearly identifiable (e.g. the first in the list, if the return value is a list).

Author: Brought up the question in today's privacy CG [meeting](https://github.com/privacycg/meetings/blob/main/2024/telcons/08-08-agenda.md), nobody brought up any initial concerns with the designs. However, I think we should ask for additional feedback once the API we finalize the language for the API to make sure.

Author: I've given it some more thought and I think that the API we need is `cookies.getPartitionKey()`, not `cookies.getHasCrossSiteAncestor()`. By getting the full key, we simplify the flow for the developer, since they will not need to create the key after calling `cookies.getHasCrossSiteAncestor()`. This also allows the browser to make the decision on what key it prefers for a given frame, if there is more than one valid one. In the event the key provided by the browser is not the one needed by the developer, the developer will have enough information from the initial call to `cookies.getPartitionKey()` to calculate the key they require.

Author: Added language to the PR to describe the API in more detail.

Author: @Rob--W, any concerns with adding this api to the proposal?

### Additional Security Considerations
None

## Alternatives
The hasCrossSiteAncestor value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the topLevelSite in the partitionKey. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for hasCrossSiteAncestor since it would not be explicitly provided by the extension.

### Existing Workarounds
To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.

### Open Web API
The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.

## Implementation Notes

### Populating hasCrossSiteAncestor when not provided

> When no value has been provided for `hasCrossSiteAncestor`, if the `domain` associated with the `cookie` is same-site to the value of the `topLevelSite`, the `hasCrossSiteAncestor` value will be set to false otherwise the value will be set to true.

Reviewer 2: The topLevelSite value is empty for unpartitioned cookies, at least in Firefox's cookies API. When topLevelSite is empty, then there is no meaningful interpretation of `if the domain associated with the cookie is same-site to the value of the topLevelSite`. What default is intended then? I think that false would make most sense as the default for unpartitioned cookies.

Here are the only possible combinations of topLevelSite and hasCrossSiteAncestor in Firefox if we were to add hasCrossSiteAncestor. Does this match Chrome? If not, can you provide a full list of valid combinations in Chrome?

- Not partitioned (top-level documents, or frames that are same-site to the whole ancestor frame tree):
   - ~`{ "topLevelSite": "", "hasCrossSiteAncestor": true }`~ (EDIT: dropped because unpartitioned cookies cannot have a cross-site ancestor bit, according to [this comment](https://github.com/w3c/webextensions/pull/581#discussion_r1704508645))
   - `{ "topLevelSite": "", "hasCrossSiteAncestor": false }`
   - (third-party frames that are not partitioned have the same options as above)
- Partitioned (third-party frames):
   - `{ "topLevelSite": "https://example.com", "hasCrossSiteAncestor": true }`
- Invalid, because a partitioned cookie (topLevelSite non-empty) implies that an ancestor was cross-site.
   - `{ "topLevelSite": "https://example.com", "hasCrossSiteAncestor": false }`

Can we include an overview of `topLevelSite` + `hasCrossSiteAncestor` pairs like this in the proposal? I think that it would be easier to follow the API with this example than the existing table for example (do keep the existing table because it serves to explain the cross-site ancestor concept).

Author: It makes sense to me that including a list of valid combinations is a good idea for our own alignment and to help developers understand the API. I've listed the combinations below for chrome. 

Valid representations of a partitionKey with hasCrossSiteAncestor included for Chrome
- Partitioned Cookie
  -  `{"topLevelSite": "https://example.com", "hasCrossSiteAncestor": false}`
      - Partitioned cookie in a same-site context, such as a partitioned cookie set on the top-level site.
  -  `{"topLevelSite": "https://example.com", "hasCrossSiteAncestor": true}`
      - Partitioned cookie in a cross-site context, such as a partitioned cookie set in a cross-site embed.
- Unpartitioned Cookie partitionKey
  - `{"topLevelSite": "", "hasCrossSiteAncestor": false}`
  - These representations will have the same behavior as the one above
    - `{}`
    - `{"topLevelSite": "", "hasCrossSiteAncestor": true}`
    - `{"hasCrossSiteAncestor": true}`
    - `{"hasCrossSiteAncestor": false}`

Reviewer 2: I see an interesting difference - the combination that I classified as invalid before appears to be valid in Chrome, as the first bullet point of your list. I suppose that if Firefox were to implement `Partitioned`, that the "invalid" case would suddenly become valid. Firefox Nightly already supports the Partitioned attribute, and I can confirm that I also see a topLevelSite in this case - TIL.

Support for top-level Partitioned means that a top-level document (and same-site child frames, potentially nested) always has access to at least TWO sets of cookies: Unpartitioned, and partitioned with `hasCrossSiteAncestor:false`. I can indeed confirm that I see two cookies when I run the following:

```javascript
document.cookie = "a=one; SameSite=None; Secure; Partitioned";
document.cookie = "a=two";
document.cookie // "a=two; a=one" in Firefox, "a=one; b=two" in Chrome.
```

Reviewer 5: Since non-URL-serialization `topLevelSites` throw, I would expect `{"topLevelSite": "", "hasCrossSiteAncestor": false}`, `{"topLevelSite": "", "hasCrossSiteAncestor": true}`, `{"hasCrossSiteAncestor": true}`, and `{"hasCrossSiteAncestor": false}` to all throw an exception.

Author: Invalid keys should result in a throw:
- `{"hasCrossSiteAncestor": false}` and `{"hasCrossSiteAncestor": true}` are invalid because the `hasCrossSiteAncestor` value is only valid within the context of the full key and there is no backward compatibility that needs to be supported.
- `{"topLevelSite": "", "hasCrossSiteAncestor": true}` is invalid because an unpartitioned cookie can't have a cross-site ancestor.

`{"topLevelSite": "", "hasCrossSiteAncestor": false}` would be considered a valid key because it reflects the key described as a valid key for an unpartitioned cookie in the background information section of the document.

Reviewer 5: Cool! That just leaves `{"topLevelSite": "", "hasCrossSiteAncestor": false}` as a gap.

To me, `{"topLevelSite": "", "hasCrossSiteAncestor": false}` is the same as `{"hasCrossSiteAncestor": false}`- neither set the full key. And[ I think the background section identifies both as valid.](https://github.com/w3c/webextensions/pull/581/files#diff-6dddb2726fe8306191f3e612a738e1b31e904af6344489bdaf93c5dbd411db80R14). 

Is there a strong reason to have this? It seems like a non-empty partitonKey object being "unpartitioned" is more confusing than helpful.

Author: I'll update the language to make it clear that a key with a `hasCrossSiteAncestor` value and no `topLevelSite` is always invalid. 

Both the Firefox and chromium implementations support the empty topLevelSite partitionKey (`{"topLevelSite": ""}`)  for this API. Since it is already supported, it made sense to me to allow developers to also pass the valid value for the `hasCrossSiteAncestor` for that key.

Reviewer 5: Ah, I misunderstood the existing behavior and thought `{"topLevelSite": ""}` would throw in exiting implementations. If that's the case, this makes sense to me!

Author: Added a table showing the valid keys with a description of the type of cookie they are associated with.

Author: @Rob--W, @bvandersloot-mozilla, gave this some more thought and removed the empty key from the table of valid partitionKeys. Since it's not used to indicate a specific partitionKey but instead used by the existing implementations of `cookies.remove()` and `cookies.getAll()` to indicate that the api should be accessing unpartitioned and partitioned cookies.

Added the information for the empty partitionKey to the implementation notes instead.

Reviewer 2: In Firefox, get/set/remove with an empty object for partitionKey is equivalent to passing `null` and means unpartitioned cookies. The remove() method removes only one cookie.

getAll() differs from the above; the object passed to getAll is treated like a filter. The use of a partitionKey property signals that the extension is aware of partitioning and passing `{}` as partitionKey causes it to return partitioned and unpartitioned cookies.

Author: Added it back to the table with the note that for `cookies.getAll()` it returns partitioned and unpartitioned cookies.

### APIs affected by the change and the behavior assoicated with the change:

- `cookies.get()`:
If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithim described above. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.

- `cookies.getAll()`: 

> If no value is set for hasCrossSiteAncestor cookies with both true and false values for hasCrossSiteAncestor will be returned. Otherwise, cookies will be returned that match the topLevelSite and the passed value for hasCrossSiteAncestor.

Reviewer 2: When I introduced the `partitionKey` property to the cookies API, I designed the API such that if the extension is unaware of the concept (due to the absence of the `partitionKey` key), that it would only get the view of unpartitioned cookies through all cookies API methods. This default made sense because top-level cookies (of the tab's primary content in the main frame) are unpartitioned. A cookie manager unaware of partitionKey would mostly work, and not inadvertently copy partitioned cookies to the unpartitioned area. For example, [EditThisCookie](https://chromewebstore.google.com/detail/editthiscookie/fngmhnnpilhplaeedifhccceomclgfbg) is a popular Chrome extension (2M users), and its current version (1.6.3) is unaware of `partitionKey`. Yet the extension mostly works, without risk of adding partitioned cookies in the mix.

In the case of `hasCrossSiteAncestor`, there is not such an immediately obvious way to separate the storages. Your proposal tries to make a reasonable guess at what the value should be in the `get` / `remove` / `set` methods, but returns all values from `cookies.getAll`. Because of this, extensions without support for `partitionKey` (let alone `hasCrossSiteAncestor`) could flip the bits and mix cookies from different partitions.

If we want to avoid this risk, a way to do so is to filter the return values of `getAll` when the extension does not pass the `partitionKey` property, and exclude cookies whose `partitionKey.hasCrossSiteAncestor` do not match the default value as provided by the algorithm. The semantics would then be as follows (modified from current MDN docs at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/getAll#partitionkey):

```diff
partitionKey Optional
An object defining which storage partitions to return cookies from:

-- if omitted, returns only cookies from unpartitioned storage.
+- if omitted, returns only cookies from unpartitioned storage and the default cross-site ancestor bit.
 - if included without topLevelSite, returns all cookies from partitioned and unpartitioned storage.
 - if included with topLevelSite specified, returns cookies from the specified partition storage.
```

Note that I did not modify the behavior when `partitionKey` is specified - the [documented](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies#storage_partitioning) recommended way to use partitionKey is to read `cookie.partitionKey` and pass its value to `cookies.set` / `cookies.remove`. Furthermore, cookies being partitioned implies them being third-party, so `hasCrossSiteAncestor` would always be true in these cases, with no risk of ambiguity.

Author: >If we want to avoid this risk, a way to do so is to filter the return values of getAll when the extension does not pass the partitionKey property, and exclude cookies whose partitionKey.hasCrossSiteAncestor do not match the default value as provided by the algorithm.

The algorithm requires both the `domain` and the `topLevelSite` to calculate the value for `hasCrossSiteAncestor`. In `cookies.getAll` if there is no `topLevelSite` (because there's no partitionKey) I'm not sure how a default value could be determined in a way that doesn't introduce a footgun. Unless this is a gap that could be covered by a new `cookies.getPartitionKey` method you [mentioned](https://github.com/w3c/webextensions/pull/581#discussion_r1703979196).

Reviewer 2: An implementation could post-process the database results and filter those where the bit differs from the default.

The Chromium implementation already has some form of post-processing, and I can imagine such check to be added easily, at https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/extensions/api/cookies/cookies_api.cc;l=385-390;drc=3f863bca976ab92d91cc73b0e17b59b1fb4c1075.

Reviewer 2: Given your response at https://github.com/w3c/webextensions/pull/581#discussion_r1704508645, the only valid hasCrossSiteAncestor value for unpartitioned cookies is `false`. Because of that, `getAll` can have the following behavior:

- If partitionKey is not set at all, only unpartitioned cookies are returned. This is the existing behavior and independent of hasCrossSiteAncestor.
- If partitionKey is set to an object, but hasCrossSiteAncestor is not specified, values with both true and false are returned. This is already what you are proposing.

I think that we can easily resolve this thread as follows:

```suggestion
If no value is set for hasCrossSiteAncestor cookies with both true and false values for hasCrossSiteAncestor will be returned. Otherwise, cookies will be returned that match the topLevelSite and the passed value for hasCrossSiteAncestor. When the `partitionKey` property is not specified, only unpartitioned cookies are returned. These cookies always have a `false` value for `hasCrossSiteAncestor`.
```

Author: SGTM, I'll commit the suggestion and resolve the conversation.

- `cookies.set()`: 

> As described the Abuse Mitigations section, this method will not allow a hasCrossSiteAncestor value of false, if the URL associated with the cookie and the topLevelSite in the partitionKey are not first-party. If this is attempted, an error will be returned. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. Additionally, if no `hasCrossSieAncestor` value is provided it will be populated using the algorithim described above.

Reviewer 2: ```suggestion
As described the Abuse Mitigations section, this method will not allow a hasCrossSiteAncestor value of false, if the URL associated with the cookie and the topLevelSite in the partitionKey are not first-party. If this is attempted, an error will be returned. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. Additionally, if no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above.
```

Reviewer 2: In today's meeting (pending merge at #666), you mentioned that Chrome does not automatically migrate "invalid" hasCrossSiteAncestor bits when the public suffix list updates, but only upon access.

While that makes sense from the perspective of the web page, and is indistinguishable to the alternative, that is not the case to extensions.

A (IMO good) reason for refusing to create "invalid" cookies is that it is likely a mistake from extension developers to create such cookies.

A hypothetical use case (which I never encountered in practice) is that enabling the extension to "create" invalid extensions would enable them to prepare for migration of cookies ahead of public suffix updates.

An actual reason where I'd prefer to not restrict `cookies.set` is that it may sometimes be the only way to really remove specific cookies (by updating the expiry of a specific cookie). Browser-specific bugs:
- https://issues.chromium.org/issues/40572551 ("chrome.cookies.remove does not account for paths and host-only cookies")
- https://bugzilla.mozilla.org/show_bug.cgi?id=1440263 ("browser.cookies.remove does not remove cookies with ../ in path, or any Path values that are not canonical")
- https://bugzilla.mozilla.org/show_bug.cgi?id=1388873 ("cookies methods (e.g. cookie.getAll) returns stale (expired) cookies")

Given the fact that the input (returned by `cookies.getAll` and `cookies.get`) can already contain invalid cookies, and the fact that there are use cases for using `cookies.set` with non-sensical `partitionKey.hasCrossSiteAncestor` values, would you be willing to consider relaxing the check, to allow `cookies.set` to be called for such cases?

From the API design point of view, I would favor validation to reject invalid cookies to prevent them from ending up in the database at all. But if the "invalid" cookie can appear in the cookie database in practice, then I think that there is also an argument in favor of letting extensions manage such cookies. There are ways to make it harder to create invalid cookies, such as only allowing existing ones to be managed and not new ones to be created. Or to require an opt-in in the form of a new permission (without install-time message).

@aselya @DCtheTall Thoughts on this?

- `cookies.remove()`:

> If no `hasCrossSieAncestor` value is provided it will be populated using the algorithm described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.

Reviewer: ```suggestion
If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above when determining the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.
```

> If no `hasCrossSieAncestor` value is provided it will be populated using the algorithm described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.

Reviewer 2: ```suggestion
If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.
```

To allow for this protection and UX flow to work, the extension would need to have the ability to set/get cookies with specific `hasCrossSiteAncestor` values. 

### Known Consumers
All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.

## Specification

### Schema

Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

```
// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
"details": {
  "partitionKey": {
    "topLevelSite": "https://example.com",
    "hasCrossSiteAncestor": true
  }
}
```
#### Cookies.GetPartitionKey()
Adds a new API, `cookies.getPartitionKey()` which retrives a valid `partitionKey` for the indicated frame.

> 

Reviewer 2: ```suggestion
This is an asynchronous function that returns a promise.
```

Fixed typo. Elsewhere "retrives" is misspelled multiple times, it should be "retrieves".

This is an asyncronus function that returns a promise.

##### Syntax
```
let key = cookies.getPartitionKey(
    details // object
)
```

> ##### Parameters

Reviewer 2: (continues thread from https://github.com/w3c/webextensions/pull/581#discussion_r1703979196)

I see two distinct use cases for the getPartitionKey method:
- to get the PartitionKey that applies to cookies with the `Partitioned` attribute in the frame. This is currently covered by the method.
- to know which partition to use when Partitioned is not specified. In Chrome the default is currently to not partition, in Firefox cookies are partitioned if third-party and SSA is not used.

To cover the second use case, we could add another option that specifies the intended scenario. An alternative is to add a method that returns whether the frame can access unpartitioned cookies.

Author: I think [hasStorageAccess](https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess) should cover the second use case already.

Reviewer 2: Since we require host permissions for the MVP, extensions can indeed run a content script to query whether storage access has been granted. Good enough!

`details` object. Information about the frame to retrive information about.
>  `tabId`
integer. The ID of the tab in which the frame is.

> `frameId`
integer. The ID of the frame in the given tab.

> ##### Return value

Reviewer 2: Could you specify:
- Does this method always return a partition key? If not, when?
- Are there any (origin) permission requirements? Cautiously we could opt for requiring host permissions for the target frame, because that permission is also required to execute a script there, which in turn could query the top level origin via `location.ancestorOrigins`. I am also willing to consider not requiring origins if we deem the return value to not be too sensitive. After all, extensions can already observe partitioned cookies for non-document subresources in frames through the cookies API.
- Are there error conditions? E.g. frame does not exist, permission denied.

Author: >Does this method always return a partition key? If not, when?

It should always return a partitionKey unless an error occurs.

>Are there any (origin) permission requirements?

In the chromium implementation, `cookies.get()`, `cookies.set()` and `cookies.remove()` all require host permissions so I think it makes sense to be consistent here as well. 

> Are there error conditions? 

Both the frame does not exist and permission denied would be an error. 
Do you have any thoughts on if a mismatch between the documentId and any tabId/frameId combination should be treated as an error?

Reviewer 2: > > Does this method always return a partition key? If not, when?
> 
> It should always return a partitionKey unless an error occurs.

SGTM.

> > Are there any (origin) permission requirements?
> 
> In the chromium implementation, `cookies.get()`, `cookies.set()` and `cookies.remove()` all require host permissions so I think it makes sense to be consistent here as well.

`cookies.get`, `set`, etc. only require host permissions for the URL where the cookie is set. Not for the partition key. An extension can therefore try to set cookies for a subresource (img, etc.) without having host permissions for the iframe. In the interest of getting to a resolution quicker, let's require host permissions for the document whose partitionkey is getting queried. If there is a compelling use case we could consider relaxing this, and that would be backwards-compatible since it is always possible to go from strict to less strict.

> > Are there error conditions?
> 
> Both the frame does not exist and permission denied would be an error. Do you have any thoughts on if a mismatch between the documentId and any tabId/frameId combination should be treated as an error?

Yes, this should be an error. When extensions use all combinations to describe a target, and there is no target matching it, then it should be an error.

Reviewer 2: This thread was marked as resolved, but the updated PR does not mention anything about permission requirements or errors. Did you forget to push a commit with the updated text?

Author: Yes, pushed the commit

To allow for this protection and UX flow to work, the extension would need to have the ability to set/get cookies with specific `hasCrossSiteAncestor` values. 

### Known Consumers
All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.

## Specification

### Schema

Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

```
// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
"details": {
  "partitionKey": {
    "topLevelSite": "https://example.com",
    "hasCrossSiteAncestor": true
  }
}
```
#### Cookies.GetPartitionKey()
Adds a new API, `cookies.getPartitionKey()` which retrives a valid `partitionKey` for the indicated frame.

This is an asyncronus function that returns a promise.

##### Syntax
```

> let key = cookies.getPartitionKey(

Reviewer 2: ```suggestion
let key = await browser.cookies.getPartitionKey(
```

The API should be async.

    details // object
)
```
##### Parameters
`details` object. Information about the frame to retrive information about.
>  `tabId`
integer. The ID of the tab in which the frame is.

> > `frameId`

Reviewer 2: Let's also add `documentId`. The minimally required option combinations is as follows:

- frameId (non-zero)
- tabId + frameId (zero frameId is also permitted)
- tabId (without frameId implies frameId 0, which is the top-level frame)
- documentId (tabId/frameId are not required but permitted)

Devlin confirmed the desirability for a different API at https://github.com/w3c/webextensions/issues/91#issuecomment-2010569435 . documentId is necessary to avoid TOCTOU issues (explained in https://github.com/w3c/webextensions/issues/8).

Author: These combination seem good to me. 

Thank you for including the additional context behind adding the documentId.

Author: Added a table to clearly describe the required inputs.

### Additional Security Considerations
None

## Alternatives
The `hasCrossSiteAncestor` value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the `topLevelSite` in the `partitionKey`. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for `hasCrossSiteAncestor` since it would not be explicitly provided by the extension.

### Existing Workarounds
To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.

### Open Web API
The APIs being expanded to include the `hasCrossSiteAncestor` boolean are specific to extensions.

## Implementation Notes

### Populating hasCrossSiteAncestor when not provided:

> When no value has been provided for `hasCrossSiteAncestor`, if the `domain` associated with the `cookie` is same-site to the value of the `topLevelSite`, the `hasCrossSiteAncestor` value will be set to false otherwise the value will be set to true.

Reviewer 5: Do you mean the cookie's domain or the cookie's URL's host part? I think you mean the latter, and this is an easy mixup to make. This makes life better here because `url` is a required member of `details` for get, set, and remove.

Author: Updated language to reflect the domain of the cookie's url

### Empty PartitionKey: `{}`
In `cookies.getAll()` an empty key `partitionKey = {}` will return both unpartitioned and partitioned cookies and `cookies.remove()` will remove both unpartitioned and partitioned cookies. For `cookies.get()` and `cookies.set()` an empty partitionKey will result in an error. 

### APIs affected by the change and the behavior assoicated with the change:

- `cookies.get()`:
If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithim described above. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.

- `cookies.getAll()`: 
If no value is set for hasCrossSiteAncestor cookies with both true and false values for `hasCrossSiteAncestor` will be returned. Otherwise, cookies will be returned that match the `topLevelSite` and the passed value for `hasCrossSiteAncestor`. When the `partitionKey` property is not specified, only unpartitioned cookies are returned. These cookies always have a `false` value for `hasCrossSiteAncestor`. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.

- `cookies.set()`: 

> As described the Abuse Mitigations section, this method will not allow a `hasCrossSiteAncestor` value of false, if the URL associated with the cookie and the `topLevelSite` in the `partitionKey` are not first-party. If this is attempted, an error will be returned. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. Additionally, if no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above.

Reviewer 5: ```suggestion
As described the Abuse Mitigations section, this method will not allow a `hasCrossSiteAncestor` value of false if the URL associated with the cookie and the `topLevelSite` in the `partitionKey` are not first-party. If this is attempted, an error will be returned. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. Additionally, if no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above.
```

|`{topLevelSite : "https://example.com", hasCrossSiteAncestor: true}`| Partitioned cookie, set in a cross-site context.|
|`{topLevelSite : "https://example.com"}`| Partitioned cookie, `hasCrossSiteAncestor` will be calculated by the browser.|
|`{topLevelSite : "",  hasCrossSiteAncestor: false}`| Unpartitioned cookie.|
|`{topLevelSite : ""}`| Unpartitioned cookie, `hasCrossSiteAncestor` will default to false.|
| `{}`| Unpartitioned cookie,  `hasCrossSiteAncestor` will default to false. Except in `cookies.getAll()` where will result in both partitioned and unpartitioned cookies being returned.| 

Note: In the table below, sites A1, A2 and A3 are all first-party to each other.
| Site frame tree |Site cookie is set on| hasCrossSiteAncestor value of cookie| Nodes that can't access the cookie|
|---|---|--|--|
| A1         |A1| false         | |
| A1->A2      |A2 | false          | 
| A1->A2->A3   |A3   | false          |  
| A1->B       |B| true          | B|
| A1->B->A2 |A2   | true          | A1 and B

**Document Metadata**

**Author:** [aselya](https://github.com/aselya) 

**Sponsoring Browser:** Chrome

> **Contributors:** [DCtheTall](https://github.com/Dcthetall)

Reviewer 2: ```suggestion
**Contributors:** [DCtheTall](https://github.com/DCtheTall), [Rob--W](https://github.com/Rob--W)
```

Fixed capitalization of existing URL and added my name.

To allow for this protection and UX flow to work, the extension would need to have the ability to set/get cookies with specific `hasCrossSiteAncestor` values. 

### Known Consumers
All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.

## Specification

### Schema

Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

```
// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
"details": {
  "partitionKey": {
    "topLevelSite": "https://example.com",
    "hasCrossSiteAncestor": true
  }
}
```

> #### Cookies.GetPartitionKey()

Reviewer 2: ```suggestion
#### Cookies.getPartitionKey()
```

- `cookies.getAll()`: 
If no value is set for hasCrossSiteAncestor cookies with both true and false values for `hasCrossSiteAncestor` will be returned. Otherwise, cookies will be returned that match the `topLevelSite` and the passed value for `hasCrossSiteAncestor`. When the `partitionKey` property is not specified, only unpartitioned cookies are returned. These cookies always have a `false` value for `hasCrossSiteAncestor`. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.

- `cookies.set()`: 
As described the Abuse Mitigations section, this method will not allow a `hasCrossSiteAncestor` value of false if the URL associated with the cookie and the `topLevelSite` in the `partitionKey` are not first-party. If this is attempted, an error will be returned. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. Additionally, if no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above.

- `cookies.remove()`:

> If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be urned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.

Reviewer 2: ```suggestion
If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithm described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.
```

# Proposal: Update Extensions to interact with Partitioned Cookies containing Cross-Site Ancestor Chain Bit values

**Summary**

> Allow extensions to utilize a hasCrossSiteAncestor boolean value when interacting with partitioned cookies that include a cross-site ancestor chain bit in their partiton key.

Reviewer 2: Is there any summary or description of what the cross-site ancestor bit is? This document presumes knowledge of that, but it is not obvious what it is. The linked pricacywg thread has a long discussion, but that is no good substitute for a concise description of what it is.

Can you add a description somewhere (or link to an existing one) that explains what the bit is, and what valid combinations of the bit and other cookie properties are? This is very relevant for the part of the API design that prescribes the behavior when "weird" combinations of the bit and url/topLevelSite are passed.

Author: Added a background section with a table that shows the behavior to my latest commit.

**Document Metadata**

**Author:**  aselya 

**Sponsoring Browser:** Chrome

**Contributors:** DCtheTall

**Created:** 2024-04-01

> **Related Issues:** [How do Partitioned cookies interact with browser extensions?](https://github.com/privacycg/CHIPS/issues/6)

Reviewer 2: ```suggestion
**Related Issues:** [PrivacyCG/CHIPS issue 6 - How do Partitioned cookies interact with browser extensions?](https://github.com/privacycg/CHIPS/issues/6)
```

In markdown the link target is obvious, but when rendered the context is not obvious from the label. Suggesting to include privacycg/chips in the label text.

## Motivation

> [Chrome](https://github.com/privacycg/CHIPS/issues/40) is adding a cross-site ancestor bit value to partitioned cookies. 

Reviewer 2: Not just Chrome, also Firefox. How about just linking the issue without label?

```suggestion
https://github.com/privacycg/CHIPS/issues/40 is adding a cross-site ancestor bit value to partitioned cookies.
```

### Objective

> To interact with partitioned cookies containing a cross-site ancestor chain bit correctly, extensions will need to have the ability to to specify a value (hasCrossSiteAncestor) that cooresponds to the value of the cross-site ancestor chain bit in partitioned cookies.

Reviewer 2: ```suggestion
To interact with partitioned cookies containing a cross-site ancestor chain bit correctly, extensions will need to have the ability to to specify a value (hasCrossSiteAncestor) that corresponds to the value of the cross-site ancestor chain bit in partitioned cookies.
```

> To interact with partitioned cookies containing a cross-site ancestor chain bit correctly, extensions will need to have the ability to to specify a value (hasCrossSiteAncestor) that cooresponds to the value of the cross-site ancestor chain bit in partitioned cookies.

Reviewer: If I'm understanding correctly, this bit makes it no longer possible to enumerate all cookies for a given origin (since there isn't a way to discover all of the possible first-party sets that could exist).

Am I understanding correctly and if so, is that something we are concerned about? It would be good to address that and either propose a new API method to resolve this or otherwise add a brief note on "non-goals" / issues we aren't attempting to solve. I don't have strong feelings either way.

Author: The ability to enumerate all cookies for a given origin should still be possible using `cookies.GetAll()` 

If a developer wants to access both partitioned and unpartitioned cookies:
A developer can pass a partitionKey without a topLevelSite and hasCrossSiteAncestor values to the `cookies.GetAll()` function and it will retrieve all cookies from a single cookie store that match the other information contained in the details object.

If a developer wants to access partitioned cookies regardless of the hasAncestorChainBit value:
A developer can pass a partitionKey with a topLevelSite but no hasCrossSiteAncestor value to the `cookies.GetAll()` function. That call will return all partitioned cookies that have the same topLevelSite and match the other information contained in the details object.

@oliverdunk, let me know if this clarification doesn't address your concern.

Reviewer: Thanks for taking the time to explain!

> If a developer wants to access partitioned cookies regardless of the hasAncestorChainBit value: A developer can pass a partitionKey with a topLevelSite but no hasCrossSiteAncestor value to the cookies.GetAll() function.

Ah, this makes sense. I missed that the behavior for cookies.get and cookies.getAll were different and getAll does not have the same logic to try and deduce a value for `hasCrossSiteAncestor`.

Author: No problem at all, glad glad to clarify.

#### Use Cases

> The methods used to get, set, remove and getAll cookies.

Reviewer 2: This is not a use case. In this section here, I would expect an explanation of what the proposed API can be used for.

For an example, see this earlier public draft that a Chrome engineer shared with me, before the partitionKey property was added to the cookies API in Chrome: https://docs.google.com/document/d/1iZKjdUft0x3m6pTH8hnppF5VHinCy0Wlprk7bUwV-XY/edit

Author: Added a new commit with a use case that describes what the proposed API can be used for.

### Known Consumers

> All adopters of partitioned cookies

Reviewer 2: The current description is vague. Here is a suggestion for a more detailed description. Does this match your intended message?

```suggestion
All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.
```

## Specification

### Schema

> Adds a new optional boolean value hasCrossSiteAncestor to the partitionKey of the cookies details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.

Reviewer 2: ```suggestion
Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.

Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.
```

### Exposed Sensitive Data
No sensitive data is associated with the cross-site ancestor chain bit.

### Abuse Mitigations
A cookie may only have no cross-site ancestor when the topLevelSite in the partitionKey and the URL with which the cookie is associated with are first party to each other. To prevent the creation of cookies that violate this, the set method will return an error if the URL and the topLevelSite are not first party for cookies that are set with no-cross site ancestor.

### Additional Security Considerations
None

## Alternatives
The hasCrossSiteAncestor value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the topLevelSite in the partitionKey. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for hasCrossSiteAncestor since it would not be explicitly provided by the extension.

### Existing Workarounds

> To access cookies with the same topLevelSite but different hasCrossSiteAncestor values (A1->B->A2 context), developers can remove the partitioned attribute from the cookie and use the Storage Access Api. 

Reviewer 2: ```suggestion
To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.
```

### Open Web API
The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.

## Implementation Notes
APIs affected by the change and the behavior assoicated with the change:

- `cookies.get()`:

> If there is a partitionKey present and it has a value for topLevelSite but no value for hasCrossSiteAncestor, the get method will try to deduce what the value for hasCrossSiteAncestor is likely to be when getting the cookie. In the event that the value is incorrect, developers can either pass the correct value for hasCrossSiteAncestor or use cookies.getAll() with no hasCrossSiteAncestor value set.

Reviewer 2: The "trying to deduce" part of this logic is underspecified. A simpler API specification would be to specify the behavior when the property or value is omitted, which I did in a sketch of an API at the end of my comment at https://github.com/privacycg/CHIPS/issues/6#issuecomment-1969920812

There I specified the behavior for get/set/remove (together) and getAll separately. The reason for this is that get/set/remove are methods to effectively select one cookie, whereas getAll queries any number of cookies. In addition to the basic query logic, the `set` method can have additional validation (e.g. to prevent non-sensical cookies to be created).

Can you specify the precise logic to determine the default, and if not, could a simpler logic be used instead (i.e. defaulting to false)?

I would prefer a simple fixed default over magical behavior, but I am willing to consider alternatives if specified well.

Author: Does the following language work for you?

In `cookies.get()` if a partitionKey is present and it has a value for `topLevelSite` but no value for `hasCrossSiteAncestor` the following algorithm will be used to calculate the value for the `hasCrossSiteAncestor`:

If a `topLevelSite` and the `url` associated with the cookie are the same, then the `hasCrossSiteAncestor` value will be set to `false` otherwise the value should default to `true`.

Reviewer 2: > Does the following language work for you?
> 
> If a `topLevelSite` and the `url` associated with the cookie are the same, then the `hasCrossSiteAncestor` value will be set to `false` otherwise the value should default to `true`.

Why do you want to have dynamic behavior, instead of always defaulting to false?

From the given examples, it looks like the "cross site ancestor bit" is part of the frame tree.

When URL and topLevelSite are cross-site, then the cookie is cross-site. But the frame in which it is loaded can still be same-site across the whole ancestor chain, i.e. `hasCrossSiteAncestor: false`

Author: If the `url` and `topLevelSite` aren't the same, the the default value of `false` for the hasCrossSiteAncestor value would be invalid. In the scenario referenced, since no value is provided for `hasCrossSiteAncestor`, the API should not default to an invalid value.

Additionally, ensuring that the value for `hasCrossSiteAncestor` can not be invalid, aligns it with the behavior present in `cookies.set()` which performs a similar check.

Reviewer 3: > The "trying to deduce" part of this logic is underspecified.

+1. I think this can be written a bit more procedurally. It is up to us to determine what "trying to deduce" means here.

> Why do you want to have dynamic behavior, instead of always defaulting to false?

Because, at least for our implementation in Chrome, most cases partitioned cookies actually have this value set to `true` unless they are set by the first-party site without any cross-site ancestors. So really there is only a special case where this is not `true`, hence I think that is probably the more sensible default.

That being said, as long as the `url` and `topLevelSite` are cross-site, then `hasCrossSiteAncestor` should *never* be false. I actually think it might even be wise to throw an error if a user tries to pass `false` in this case.

On the other hand, if `url` and `topLevelSite` are same-site, then we could have cookies whose key has a `false` `hasCrossSiteAncestor` and cookies whose keys have a `true` value (those set in A>B>A contexts). I think in the `url` and `topLevelSite` being same-site case it is reasonable to default to `false` for `set`, `get`, and `delete`. But users can provide a value manually if they wish to access cookies in A>B>A contexts.

For `getAll` I think it is reasonable to include both values (when it makes sense) in our query for cookies if a value is not provided.

How does that all sound?

Reviewer 2: > > Why do you want to have dynamic behavior, instead of always defaulting to false?
> 
> 
> 
> Because, at least for our implementation in Chrome, most cases partitioned cookies actually have this value set to `true` unless they are set by the first-party site without any cross-site ancestors. So really there is only a special case where this is not `true`, hence I think that is probably the more sensible default.

Are you advocating for unconditional true, or for automatic deduction?

> That being said, as long as the `url` and `topLevelSite` are cross-site, then `hasCrossSiteAncestor` should *never* be false. I actually think it might even be wise to throw an error if a user tries to pass `false` in this case.

Whether two URLs are same-site is not fixed. When the Public Suffix List (PSL) updates, the same-site-ness of two domains may change. Allowing false enables cookie managers to migrate cookies in this case.

> How does that all sound?

Other than what I commented on, sgtm.

Author: > > > Why do you want to have dynamic behavior, instead of always defaulting to false?
> > 
> > 
> > Because, at least for our implementation in Chrome, most cases partitioned cookies actually have this value set to `true` unless they are set by the first-party site without any cross-site ancestors. So really there is only a special case where this is not `true`, hence I think that is probably the more sensible default.
> 
> Are you advocating for unconditional true, or for automatic deduction?
> 
> > That being said, as long as the `url` and `topLevelSite` are cross-site, then `hasCrossSiteAncestor` should _never_ be false. I actually think it might even be wise to throw an error if a user tries to pass `false` in this case.
>>> Having an invalid `hasCrossSiteAncestor` value throw an error makes sense to me. When the value is not provided, it should be automatically calculated. Since defaulting to `true` or `false` without evaluating if those values could be valid, will result in cookies with invalid partition keys. 

> Whether two URLs are same-site is not fixed. When the Public Suffix List (PSL) updates, the same-site-ness of two domains may change. Allowing false enables cookie managers to migrate cookies in this case.
>> To handle this use case, a cookie manager looking to do a migration, would be able to use `getAll()` to obtain all the cookies present (regardless of if their partition keys are valid after a PSL update).  Then the cookie manager could iterate through the cookies and use `set()` to attempt to create cookies with the `hasCrossSiteAncestor` value of `false`. If the `set()` returns an error, the manager will know the `topLevelSite` and the `url` are still cross-site after the PSL update. 
> > How does that all sound?
> 
> Other than what I commented on, sgtm.

Author: Resolving because discussion is continuing in more recent [comment].(https://github.com/w3c/webextensions/pull/581#discussion_r1701585197)

- `cookies.getAll()`: 
If no value is set for hasCrossSiteAncestor cookies with both true and false values for hasCrossSiteAncestor will be returned. Otherwise, cookies will be returned that match the topLevelSite and the passed value for hasCrossSiteAncestor.

- `cookies.set()`: 
As described the Abuse Mitigations section, this method will not allow a hasCrossSiteAncestor value of false, if the URL associated with the cookie and the topLevelSite in the partitionKey are not first party. If this is attempted, an error will be returned. If no topLevelSite is provided and a hasCrossSiteAncestor value is provided, the cookie will not be set and an error will be returned.

- `cookies.remove()`:

> If no value is set for hasCrossSiteAncestor, cookies.remove() will remove cookies with both true and false values. If no topLevelSite value is included in the partitionKey object and a value is set for hasCrossSiteAncestor, no cookies will be removed and an error will be returned.

Reviewer 2: The proposal here explicitly specifies that multiple cookies ought to be matched. But it is unclear whether the `cookies.remove` should really be removing all cookies or just the first match.

In Firefox, the cookies.get/set/remove methods all share the same cookie selection logic.

Should `cookies.remove` match and replace multiple cookies by design, or should it match one (and the removeAll method be introduced to allow fuzzy removal of multiple cookies)? Relevant issues:

- https://issues.chromium.org/issues/40572551
- https://bugzilla.mozilla.org/show_bug.cgi?id=1387957

Author: Updated language to remove reference to deleting multiple cookies.

If the logic for `cookies.remove()` is to be changed, I think it's beyond the scope of this PR.

As described the Abuse Mitigations section, this method will not allow a hasCrossSiteAncestor value of false, if the URL associated with the cookie and the topLevelSite in the partitionKey are not first-party. If this is attempted, an error will be returned. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. Additionally, if no `hasCrossSieAncestor` value is provided it will be populated using the algorithim described above.

- `cookies.remove()`:

> If no `hasCrossSieAncestor` value is provided it will be populated using the algorithim described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.

Reviewer 4: ```suggestion
If no `hasCrossSieAncestor` value is provided it will be populated using the algorithm described above when determing the cookie to remove. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned. If `topLevelSite` and `hasCrossSiteAncestor` values are provided, they will be used by the method even if the combination of the values would be invalid.
```
